Blocks: Propose a server side rendering implementation

youknowriad opened the pull request:
This is also called dynamic blocks and is useful for blocks like embed, comments blocks, ...
This is still in progress but feedback is appreciated.

**Testing instructions**

 - Insert this `<!-- wp:core/html -->anything<!-- /wp:core/html -->` to a post content (old editor) and preview this post, you should see your block replaced by `THIS IS AWESOME`

youknowriad committed:
Blocks: Propose a server side rendering implementation

This is also called dynamic blocks and is usefull for blocks like embed, comments blocks, ...

aduth commented on index.php:
```diff
@@ -27,6 +27,75 @@ function gutenberg_menu() {
+ */
+function register_block( $name, $render ) {
+	global $registered_blocks;
+	$registered_blocks[ $name ] = $render;
```
I don't want to preemptively assume requirements, but do you think we'd need a block to register options aside from just a render?

youknowriad replied:
I tried to think about this, but I didn't find any use-case for now. Though, we don't know the future maybe options is preferable.

aduth commented on index.php:
```diff
@@ -27,6 +27,75 @@ function gutenberg_menu() {
+
+ * @return string          Updated post content
+ */
+function do_dynamic_blocks( $content ) {
```
This function could do for some comments; various indices are hard to follow.

mtias replied:
Is the specificity of "dynamic" something we want to expose in the function name? I'd expect we'll add hooks here for interacting with individual blocks (say I want to filter all quote blocks for some reason). Performance here will be very important as well.

aduth commented on index.php:
```diff
@@ -27,6 +27,75 @@ function gutenberg_menu() {
+	$new_content = $content;
+	foreach ( $matches[ 0 ] as $index => $block_match ) {
+		$block_name = $matches[ 1 ][ $index ][ 0 ];
+		if ( isset( $registered_blocks[ $block_name ] ) ) {
```
Maybe worth an early continue to avoid indentation:

```php
if ( ! isset( $registered_blocks[ $block_name ] ) ) {
	continue;
}
```

aduth commented on index.php:
```diff
@@ -27,6 +27,75 @@ function gutenberg_menu() {
+	foreach ( $matches[ 0 ] as $index => $block_match ) {
+		$block_name = $matches[ 1 ][ $index ][ 0 ];
+		if ( isset( $registered_blocks[ $block_name ] ) ) {
+			$block_length = mb_strlen( $block_match[ 0 ] );
```
I don't know the answer, but is `mb_strlen` the safe default over `strlen` for multi-byte content?

youknowriad replied:
In my PHP souvenirs, it is, but I'm not in my comfort zone anymore in PHP. Other 👀 here could be useful.

aduth commented on index.php:
```diff
@@ -247,3 +316,7 @@ function the_gutenberg_project() {
 	<?php
 }
+
+register_block( 'core/html', function( $attributes ) {
```
Seems temporary, but just for clarity, the anonymous function callable is only available in PHP versions 5.3.0 and newer ([reference](http://php.net/manual/en/functions.anonymous.php)). WordPress supports as far back as 5.2.4.

youknowriad replied:
Oh right 👍

aduth commented on index.php:
```diff
@@ -27,6 +27,75 @@ function gutenberg_menu() {
+ */
+function do_dynamic_blocks( $content ) {
+	global $registered_blocks;
+	$open_matcher = '/<!--\s*wp:([a-z](?:[a-z0-9\/]+)*)\s+((?:(?!-->).)*)-->.*<!--\s*\/wp:\g1\s+-->/';
```
Would we want the registered block to be able to make use of the content between the comment delimiters?

youknowriad replied:
I know there were some discussions about this, but IMO for consistency, Dynamic blocks should not contain any data between the delimiters. Their `save` function should return `null`.

mtias replied:
We shouldn't prescribe what save returns, because it may make sense to return a link, or a short message. Consider a block that renders the last 5 recent posts from a certain category. It may be good that when the dynamic markup can't be generated, it returns a link to `/blog/category/:term` with some text saying "View latest posts on :term".

mtias commented:
Thanks for starting this.

youknowriad committed:
Blocks: Registering a block server side takes an array

 - Also, ensure PHP 5.2 compatibility
 - Adding some comments to explain the server side rendering

youknowriad commented:
I'm thinking to leverage this for the `embed` block. But to do so, we need to change the way we store attributes (use HTML like syntax). 

What do you think if 
 - we remove the dummy "register_block" call. 
 - Try to merge this proposal
 - do a follow-up PR to change the attributes' syntax 
 - and then another PR to add the server side embed block. 

(I like splitting/Iterating :) )

aduth commented:
I'm not totally clear what you mean by changing attribute syntax, but I'd be fine with merging this and iterating on it in subsequent pull requests since it has no impact on existing behavior for the moment.

nylen commented:
This logic looks fairly complicated and needs unit tests as soon as possible.  See https://pippinsplugins.com/unit-tests-wordpress-plugins-setting-up-testing-suite/ for example.

I think we should do more to unify our block registration APIs between the client and server (require the same rules for block slugs, for example).  I also wonder if we need or want some way to guarantee that if a block is registered on the server, then it's also registered on the client (and the relevant details match up).

I'm not sure those need to be done in this PR, but I'm also not sure why we need to merge this immediately.  Are we ready to use it yet?

BE-Webdesign commented:
What does the server side API need to do and why does it need to do it, would be good questions to answer before we get too attached to this. I am fine with merging, but I think the original JS API could use improvements right now as well, and we should circle back to refining the API at some point.

westonruter commented:
Additionally, I think there is also the question as to whether the server-side API should also be able to facilitate rendering of blocks for displaying inside of the visual editor, similar to what the Shortcake plugin does. A good case for server-side rendering for the block in the editor is both for legacy content (e.g. shortcodes) but also for displaying dynamic blocks that are not feasible to provide pure JS renderings without duplicating a lot of logic.

youknowriad commented:
@aduth 

> I'm not totally clear what you mean by changing attribute syntax

moving from `attr:value` to `attr="value"` or `{ attr: "value" }`. We can't store URLs for example for now.

@nylen 

> This logic looks fairly complicated

I don't think so

> needs unit tests as soon as possible

I agree

> I think we should do more to unify our block registration APIs between the client and server

Sounds good for the validation

> . I also wonder if we need or want some way to guarantee that if a block is registered on the server, then it's also registered on the client

I don't think so, we only need the dynamic blocks on the server, why forcing plugin authors to add something optional

>  but I'm also not sure why we need to merge this immediately

I'm not so impatient to merge this, I was waiting for review and you raise some good points here. But  I like to split the work to smaller PRs.

> Are we ready to use it yet?

Yes we are, we can use this to generate the HTML for the embed block for example.


@BE-Webdesign 

> What does the server side API need to do, why does it need to do it

The only thing that it needs to do for now is rendering dynamic blocks and we currently have a dynamic block on the client side (The embed block). I'm trying to start the server side implementation to tackle this particular task. We'll iterate and see if we need any other server side API.

@westonruter 

> Additionally, I think there is also the question as to whether the server-side API should also be able to facilitate rendering of blocks for displaying inside of the visual editor

I'm not sure I understand! Are you saying we should generate a form in the visual editor server-side? I would personally strongly argue against that and instead, build a specific "nice" UI for each block. But I understand this could be helpful for mobile apps maybe. Anyway, this is a separate task and for now, I'm trying to address the server-side rendering of dynamic blocks.

youknowriad committed:
Blocks: Validate the registred block server side

youknowriad committed:
SSR: Simplify the block server side rendering algorithm

nylen commented:
> I'm not so impatient to merge this, I was waiting for review and you raise some good points here. But I like to split the work to smaller PRs.

I agree, I guess it makes me a bit nervous that we are building a copy of an API that already exists on the client side.  We need to be careful that it works the same way as much as possible.  To me that means we should follow a similar process here to #289, including the same test cases.  So I'd prefer that we finish and merge #617 (PHPUnit) first, then add test cases here.

youknowriad committed:
Setup server side unit tests (#617)

* Setup server side unit tests

youknowriad commented:
Server side tests in place, any other concern here?

nylen commented:
This was also present on #617, but I'm seeing a lot of warnings in the build log:

> PHP Warning:  filemtime(): stat failed for /home/travis/build/WordPress/gutenberg/i18n/build/index.js in /home/travis/build/WordPress/gutenberg/index.php on line 161

Example: https://travis-ci.org/WordPress/gutenberg/jobs/228752501

Also, from https://github.com/WordPress/gutenberg/pull/617#issuecomment-299185065, was this ever addressed?

> Take a close look at each of the jobs, there are two showing "green" that they passed, this is false, there's lots of errors on those jobs.

Let's fix the warnings here and make sure that when the build fails it actually fails like it should.  I'll do a separate PR against this branch to test out the second thing.

youknowriad commented:
Oh! I haven't seen those errors. I thought he was talking about the failing jobs. At that moment three of five jobs were failing.

nylen commented on phpunit/class-dynamic-blocks-render-test.php:
```diff
@@ -0,0 +1,51 @@
+	/**
+	 * Successfull dynamic block rendering
+	 */
+	function test_register_block() {
```
In this PR there have been issues with matching the content inside the block delimiters. Let's get another test similar to this one but testing out content inside blocks as well:

```html
before
<!-- wp:core/dummy value:a1 -->this should be ignored<!-- /wp:core/dummy -->
between
<!-- wp:core/dummy value:a2 -->this should be ignored also<!-- /wp:core/dummy -->
after
```

nylen commented on index.php:
```diff
@@ -27,6 +27,121 @@ function gutenberg_menu() {
+		return;
+	}
+
+	if ( isset( $registered_blocks[ $slug ] ) ) {
```
Why not trigger a `_doing_it_wrong` here too, like we do on the client side?

nylen commented on index.php:
```diff
@@ -27,6 +27,121 @@ function gutenberg_menu() {
+ */
+function unregister_block( $slug ) {
+	global $registered_blocks;
+	if ( ! isset( $registered_blocks[ $slug ] ) ) {
```
Why not trigger a `_doing_it_wrong` here if this is set, like we do on the client side?

nylen commented:
Here are the block registration test cases on the client side:

```
  blocks
    registerBlock()
      ✓ should reject numbers
      ✓ should reject blocks without a namespace
      ✓ should reject blocks with invalid characters
      ✓ should accept valid block names
      ✓ should prohibit registering the same block twice
      ✓ should store a copy of block settings
    unregisterBlock()
      ✓ should fail if a block is not registered
      ✓ should unregister existing blocks
```

Let's get all of the same tests running here too.

nylen committed:
Apply Travis CI improvements from #676:

- Do the client-side build before running PHPUnit tests; this ensures
  that the plugin doesn't try to enqueue non-existent JS files and cause
  a bunch of warnings
- If a step in the build fails, go ahead and exit
- Don't scan node_modules/ and vendor/ for PHP files
- Run phpcs with the `-s` option (show rule identifiers in reports)
- Don't require docblocks for unit tests

youknowriad committed:
SSR: add missing doing it wrong

youknowriad committed:
Testing: Add more registration unit tests

youknowriad commented:
I've added more unit tests fc7fabe (closer to what we have on the frontend). Can I have another look @nylen?

nylen committed:
Prefix global name with `wp_`

nylen committed:
Use `tearDown` functions rather than unregistering blocks in each test

nylen committed:
Test that $wp_registered_blocks global is modified as expected

nylen committed:
Add a missing test for `register_block`

nylen committed:
Explicitly return false for errors and test this condition

nylen committed:
Add test with (ignored) content inside block delimiters

nylen committed:
Make it possible to run `phpunit` from within an existing WP dev install

Before this change, running the unit tests in this configuration
required an extra step (`WP_TESTS_DIR=...  phpunit`).

nylen commented:
There were still a couple of tests missing:

- should prohibit registering the same block twice
- https://github.com/WordPress/gutenberg/pull/586#discussion_r114985211
- verify that the functions actually make the expected changes to the blocks global variable

Other changes:

- As of 47e81bd every test method no longer needs to have an annotation.
- Changed `$registered_blocks` to `$wp_registered_blocks`; in WP, [prefix everything](https://nacin.com/2010/05/11/in-wordpress-prefix-everything/).
- Explicit `return false;` for error conditions; also test accordingly.
- You can now run `phpunit` from within a WP dev install without additional configuration or having to specify `WP_TESTS_DIR`.

nylen committed:
Fix a logic error

nylen committed:
Fix style in PHP comments

nylen reviewed (approved):
If latest changes look OK: :shipit:
